prov/sm2/src/sm2_progress.c

```diff
@@ -289,12 +435,98 @@ static int sm2_alloc_xfer_entry_ctx(struct sm2_ep *ep,
 static int sm2_progress_recv_msg(struct sm2_ep *ep,
 				 struct sm2_xfer_entry *xfer_entry)
 {
+	struct sm2_av *av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
 	struct fid_peer_srx *peer_srx = sm2_get_peer_srx(ep);
 	struct fi_peer_rx_entry *rx_entry;
+	struct sm2_cma_data *cma_data;
 	struct sm2_av *sm2_av;
 	fi_addr_t addr;
+	uint64_t comp_flags;
 	int ret;
 
+	/* TODO - Switch on protocol before switching on op to avoid messy
+	 * checks like this */
```

Comment:
would prefer if this TODO was addressed in this patch because this is very messy as is. maybe something like this?
```
if (xfer_entry->hdr.proto_flags & (HOST_TO_DEV | HOST_TO_DEV_ACK))
    sm2_handle_cma_ipc()

//do regular message processing
```

there's a lot in this function. I think checking the flags first is ok but I would move it out of this function.

Reply:
proto_flags are protocol specific, so we can't switch on proto_flags. We don't have enough bits to make proto_flags independent of protocol and Seth doesn't want to make the header longer.

I agree it's messy but at this point we are focusing on getting things functional and into our CI